yarn-project/archiver/src/archiver/data_retrieval.ts

```diff
@@ -126,6 +132,74 @@ export async function getL1BlockTime(publicClient: PublicClient, blockNumber: bi
   return block.timestamp;
 }
 
+/**
+ * Extracts the first 'propose' method calldata from a forwarder transaction's data.
+ * @param forwarderData - The forwarder transaction input data
+ * @param rollupAddress - The address of the rollup contract
+ * @returns The calldata for the first 'propose' method call to the rollup contract
+ */
+function extractRollupProposeCalldata(forwarderData: Hex, rollupAddress: Hex): Hex {
+  // TODO(#11451): custom forwarders
+  const { functionName: forwarderFunctionName, args: forwarderArgs } = decodeFunctionData({
+    abi: ForwarderAbi,
+    data: forwarderData,
+  });
+
+  if (forwarderFunctionName !== 'forward') {
+    throw new Error(`Unexpected forwarder method called ${forwarderFunctionName}`);
+  }
+
+  if (forwarderArgs.length !== 2) {
+    throw new Error(`Unexpected number of arguments for forwarder`);
+  }
+
+  const [to, data] = forwarderArgs;
+
+  // Find all rollup calls
+  const rollupAddressLower = rollupAddress.toLowerCase();
+  const rollupCalls = to.reduce((acc: number[], arg, index) => {
+    if (arg.toLowerCase() === rollupAddressLower) {
+      acc.push(index);
+    }
+    return acc;
+  }, []);
+
+  if (rollupCalls.length === 0) {
+    throw new Error(`Rollup address not found in forwarder args`);
+  }
+
+  // Find the first propose method among all rollup calls
+  let rollupData: Hex | undefined;
+
+  for (const index of rollupCalls) {
+    if (index >= data.length) {
+      continue;
+    }
+
+    const callData = data[index];
+    try {
+      const { functionName: rollupFunctionName } = decodeFunctionData({
+        abi: RollupAbi,
+        data: callData,
+      });
+
+      if (rollupFunctionName === 'propose') {
+        rollupData = callData;
```

Comment:
i feel like we can just return immediately here. and skip reducing the rollup calls